### DISCUSSION TITLE
How to list all files been tracked by git?

### DISCUSSION THREAD

**Discussion by shlomiLan** [gitpython-developers/GitPython#1277] [Q&A] [answered]
Hey all,

How do I get a list of all file? using git I can run 
```bash
git ls-tree HEAD -r
```

how can I do it with this library? the best I could find is:
```python
for file in repo.commit().tree.blobs:
    print(file.abspath)
```
but that only lists files in the top dictionary.

Thanks

**Comment by Byron** [comment, marked as answer]
You are probably looking for this:

```python
for entry in repo.commit().tree.traverse():
    print(entry)
```

There are fields or methods on the returned entries to obtain their name or path relative to the root tree.

**Comment by ChrisBarker-NOAA** [comment]
To complete this, if you want a file listing, you want:

```
for entry in repo.commit().tree.traverse():
    print(entry.path)
```
or 
```
for entry in repo.commit().tree.traverse():
    print(entry.abspath)
```

**Reply by Byron** [reply]
I also realize now that even though this is a way to list files that will always work, it will be much faster to traverse index entries instead if one is present, i.e. do the equivalent of `git ls-files`. But then again, unless `repo.git.ls_files()` is used, any kind of traversal will be slow comparatively.
